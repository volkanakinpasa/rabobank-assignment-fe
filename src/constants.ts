const API_URL = 'https://demo-api.now.sh';

export const LINKS = {
  SIGN_UP: '/',
};

export const API_ENDPOINTS = {
  USERS: `${API_URL}/users`,
};
