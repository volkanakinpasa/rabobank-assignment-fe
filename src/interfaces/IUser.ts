export default interface IUser {
  firstName: string;
  lastName: string;
  email: string;
  _id: string;
}
