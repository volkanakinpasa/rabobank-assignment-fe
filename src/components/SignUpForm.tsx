import './style.css';

import { FORM_INPUT_IDS, FORM_INPUT_NAMES, MESSAGES } from '../constants';
import { get, post } from '../api/users';

import ISignUpForm from '../interfaces/ISignUpForm';
import IUser from '../interfaces/IUser';
import Loading from './Loading';
import SignedUp from './SignedUp';
import { useForm } from 'react-hook-form';
import { useState } from 'react';

export default function SignUpForm() {
  const [loading, setLoading] = useState<boolean>(false);
  const [showForm, setShowForm] = useState<boolean>(true);
  const [showSignedUp, setShowSignedUp] = useState<boolean>(false);

  const {
    register,
    handleSubmit,
    formState: { errors },
    getValues,
  } = useForm<ISignUpForm>();

  const validatePassword = (password: string): boolean => {
    const { firstName, lastName } = getValues();

    //what does Should contain lower and uppercase letters mean?. should it contain numbers? or do you mean only accepts lowercase and uppercase?
    // /(?=.*[a-z])(?=.*[A-Z])/.test(password) &&

    if (
      firstName.length > 0 &&
      password.toLowerCase().indexOf(firstName.toLowerCase()) > -1
    ) {
      return false;
    } else if (
      lastName.length > 0 &&
      password.toLowerCase().indexOf(lastName.toLowerCase()) > -1
    ) {
      return false;
    } else {
      return true;
    }
  };

  const onSubmit = async (data: ISignUpForm) => {
    try {
      setLoading(true);
      const result: IUser = await post(data);
      console.log(result);

      setTimeout(async () => {
        const users: IUser[] = await get(result._id);

        console.log(users);
        setShowForm(false);
        setShowSignedUp(true);
      }, 4000);
    } catch {
      setLoading(false);
    }
  };

  const Label = ({
    name,
    labelFor,
  }: {
    name: string;
    labelFor: string;
  }): JSX.Element => <label htmlFor={labelFor}>{name}</label>;

  const { ERRORS, EMAIL_PATTERN } = MESSAGES;
  const {
    FIRST_NAME_REQUIRED,
    LAST_NAME_REQUIRED,
    EMAIL_REQUIRED,
    EMAIL_INVALID,
    PASSWORD_REQUIRED,
    PASSWORD_MIN,
    PASSWORD_VALIDATE,
  } = ERRORS;

  const { FIRST_NAME, LAST_NAME, EMAIL, PASSWORD } = FORM_INPUT_NAMES;

  return (
    <div className="signup-container">
      {showForm && (
        <form onSubmit={handleSubmit(onSubmit)}>
          <Label name={FIRST_NAME} labelFor={FORM_INPUT_IDS.FIRST_NAME} />
          <input
            id={FORM_INPUT_IDS.FIRST_NAME}
            type="text"
            placeholder={FIRST_NAME}
            autoComplete="name"
            {...register(`firstName`, {
              required: FIRST_NAME_REQUIRED,
            })}
          />
          {errors.firstName && (
            <div>
              <p>{errors.firstName.message}</p>
            </div>
          )}

          <Label name={LAST_NAME} labelFor={FORM_INPUT_IDS.LAST_NAME} />
          <input
            id={FORM_INPUT_IDS.LAST_NAME}
            type="text"
            placeholder={LAST_NAME}
            {...register('lastName', {
              required: LAST_NAME_REQUIRED,
            })}
          />
          {errors.lastName && (
            <div>
              <p>{errors.lastName.message}</p>
            </div>
          )}
          <Label name={EMAIL} labelFor={FORM_INPUT_IDS.EMAIL} />
          <input
            id={FORM_INPUT_IDS.EMAIL}
            type="text"
            autoComplete={FORM_INPUT_IDS.EMAIL}
            placeholder={EMAIL}
            {...register('email', {
              required: EMAIL_REQUIRED,
              pattern: {
                value: EMAIL_PATTERN,
                message: EMAIL_INVALID,
              },
            })}
          />
          {errors.email && (
            <div>
              <p>{errors.email.message}</p>
            </div>
          )}
          <Label name={PASSWORD} labelFor={FORM_INPUT_IDS.PASSWORD} />
          <input
            type="password"
            id={FORM_INPUT_IDS.PASSWORD}
            autoComplete="current-password"
            placeholder={PASSWORD}
            data-testid={FORM_INPUT_IDS.PASSWORD}
            {...register('password', {
              required: PASSWORD_REQUIRED,
              minLength: { value: 8, message: PASSWORD_MIN },
              validate: validatePassword,
            })}
          />
          {errors.password && (
            <div>
              <p>{errors.password.message}</p>
              {errors.password.type === 'validate' && (
                <p>{PASSWORD_VALIDATE}</p>
              )}
            </div>
          )}

          {!loading && (
            <button type="submit" color="primary">
              Sign Up
            </button>
          )}

          {loading && <Loading />}
        </form>
      )}
      {showSignedUp && <SignedUp />}
    </div>
  );
}
