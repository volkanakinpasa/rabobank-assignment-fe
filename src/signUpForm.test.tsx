import { render, screen, waitFor } from '@testing-library/react';

import IUser from './interfaces/IUser';
import { MESSAGES } from './constants';
import SignUpForm from './components/SignUpForm';
import userEvent from '@testing-library/user-event';

const { ERRORS } = MESSAGES;
const {
  FIRST_NAME_REQUIRED,
  LAST_NAME_REQUIRED,
  EMAIL_REQUIRED,
  EMAIL_INVALID,
  PASSWORD_REQUIRED,
  PASSWORD_MIN,
  PASSWORD_VALIDATE,
} = ERRORS;

jest.mock('./api/users', () => ({
  post: () =>
    new Promise<IUser>((res, rej) => {
      res({
        firstName: 'string',
        lastName: 'string',
        email: 'string',
        _id: 'string',
      });
    }),
  get: () =>
    new Promise<IUser[]>((res, rej) => {
      res([
        {
          firstName: 'string',
          lastName: 'string',
          email: 'string',
          _id: 'string',
        },
      ]);
    }),
}));

jest.setTimeout(60000);

describe('SignUpForm component', () => {
  it('creates form snapshot', () => {
    const { asFragment } = render(<SignUpForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows required error messages when form submitted', async () => {
    render(<SignUpForm />);

    userEvent.click(screen.getByText(/Sign Up/i));
    await waitFor(() => {
      expect(screen.getByText(FIRST_NAME_REQUIRED)).toBeInTheDocument();
      expect(screen.getByText(LAST_NAME_REQUIRED)).toBeInTheDocument();
      expect(screen.getByText(EMAIL_REQUIRED)).toBeInTheDocument();
      expect(screen.getByText(PASSWORD_REQUIRED)).toBeInTheDocument();
    });
  });

  it('shows email is invalid when email has no @', async () => {
    render(<SignUpForm />);
    userEvent.type(screen.getByRole('textbox', { name: 'Email' }), 'test');
    userEvent.click(screen.getByText(/Sign Up/i));
    await waitFor(() => {
      expect(screen.getByText(EMAIL_INVALID)).toBeInTheDocument();
    });
  });

  it('shows password min x charters error message', async () => {
    render(<SignUpForm />);
    userEvent.type(screen.getByTestId('password'), 'test');
    userEvent.click(screen.getByText(/Sign Up/i));
    await waitFor(() => {
      expect(screen.getByText(PASSWORD_MIN)).toBeInTheDocument();
    });
  });

  it('shows password error message when first name exist in', async () => {
    render(<SignUpForm />);
    userEvent.type(
      screen.getByRole('textbox', { name: 'First Name' }),
      'fname'
    );
    userEvent.type(screen.getByTestId('password'), 'fnameSDFGHJ');
    userEvent.click(screen.getByText(/Sign Up/i));
    await waitFor(() => {
      expect(screen.getByText(PASSWORD_VALIDATE)).toBeInTheDocument();
    });
  });

  it('shows password error message when last name exist in', async () => {
    render(<SignUpForm />);
    userEvent.type(screen.getByRole('textbox', { name: 'Last Name' }), 'lname');
    userEvent.type(screen.getByTestId('password'), 'lnameSDFGHJ');
    userEvent.click(screen.getByText(/Sign Up/i));
    await waitFor(() => {
      expect(screen.getByText(PASSWORD_VALIDATE)).toBeInTheDocument();
    });
  });

  it('submits form', async () => {
    render(<SignUpForm />);

    userEvent.type(
      screen.getByRole('textbox', { name: 'First Name' }),
      'firstName'
    );
    userEvent.type(
      screen.getByRole('textbox', { name: 'Last Name' }),
      'lastName'
    );
    userEvent.type(
      screen.getByRole('textbox', { name: 'Email' }),
      'test@test.com'
    );
    userEvent.type(screen.getByTestId('password'), 'qwertyuiosdf');
    userEvent.click(screen.getByText(/Sign Up/i));

    await waitFor(
      async () => {
        await timeout(4000);
        expect(screen.getByText('Signed up!')).toBeInTheDocument();
      },
      { timeout: 40000 }
    );
  });
});

function timeout(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
