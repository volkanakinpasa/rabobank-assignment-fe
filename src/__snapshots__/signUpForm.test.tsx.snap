// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm component creates form snapshot 1`] = `
<DocumentFragment>
  <div
    class="signup-container"
  >
    <form>
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        autocomplete="name"
        id="firstName"
        name="firstName"
        placeholder="First Name"
        type="text"
      />
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        type="text"
      />
      <label
        for="email"
      >
        Email
      </label>
      <input
        autocomplete="email"
        id="email"
        name="email"
        placeholder="Email"
        type="text"
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        autocomplete="current-password"
        data-testid="password"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <button
        color="primary"
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
